### ISSUE TITLE
DataReshaper

### ISSUE THREAD

**Issue by Joaopmoliveira** [google/filament#4246]
**Is your feature request related to a problem? Please describe.**
When reading the pixels from the GPU there is a stage related with reshaping the original image, i was wondering if this could be removed in specific use cases. 

**Describe the solution you'd like**
My question is if this step could be removed when rendering a standalone view with a user provided texture, thus the image layout is known and the copy into cpu could be speed up.

**Describe alternatives you've considered**
I do not totally understand the implications of what I am describing, but I assume in the Vulkan case a VKImage is created with a given layout, if this image is already in a layout which is the same as the Texture provided as a render target then it is expected that the user would provide a PixelBufferDescriptor with the same layout and thus a simple copy would suffice after the vkCmdBlitImage command.

**OS and backend**
Not specific to any device.

**Extra info**
I saw that in the backend tests (test_Blit.cpp) there is a call to a api.blit() method, is this is dangerous to turn into a public interface?

**Comment by romainguy** [comment]
Reshaping only occurs for formats allowed by our API but not natively handled by Vulkan. If you use a format that Vulkan supports natively, there shouldn’t be any overhead. Do you have a specific example?

**Comment by Joaopmoliveira** [comment]
This is more or less the code I have been using for testing. 

```c++
using closure_t = std::function<void(uint8_t* rgba, uint32_t width, uint32_t height)>;

	Engine* mEngine = Engine::create();

	View* mView = mEngine->createView();
	mView->setViewport({ 0, 0, swapchain_width, swapchain_height });

	const Viewport& vp = mView->getViewport();

	Texture* offscreenColorTexture = Texture::Builder()
		.width(vp.width).height(vp.height).levels(1)
		.usage(Texture::Usage::SAMPLEABLE | Texture::Usage::UPLOADABLE | Texture::Usage::COLOR_ATTACHMENT)
		.format(Texture::InternalFormat::RGBA8).build(*mEngine);

	RenderTarget* offscreenRenderTarget  = RenderTarget::Builder()
		.texture(RenderTarget::AttachmentPoint::COLOR, offscreenColorTexture)
		.build(*mEngine);

	Renderer* mRenderer = mEngine->createRenderer();

	Scene* mScene = mEngine->createScene();

	utils::Entity mCameraEntity = utils::EntityManager::get().create();
	Camera* mCamera = mEngine->createCamera(mCameraEntity);

	mView->setScene(mScene);
	mView->setCamera(mCamera);
	mView->setRenderTarget(offscreenRenderTarget);

	Skybox* mSkybox = Skybox::Builder().color({ 0.1, 0.125, 0.25, 1.0 }).build(*mEngine);
	mScene->setSkybox(mSkybox);
	mView->setPostProcessingEnabled(false);

	VertexBuffer* vb = VertexBuffer::Builder()
		.vertexCount(3)
		.bufferCount(1)
		.attribute(VertexAttribute::POSITION, 0, VertexBuffer::AttributeType::FLOAT2, 0, 12)
		.attribute(VertexAttribute::COLOR, 0, VertexBuffer::AttributeType::UBYTE4, 8, 12)
		.normalized(VertexAttribute::COLOR)
		.build(*mEngine);

	vb->setBufferAt(*mEngine, 0,
		VertexBuffer::BufferDescriptor(TRIANGLE_VERTICES, 36, nullptr));

	IndexBuffer* ib = IndexBuffer::Builder()
		.indexCount(3)
		.bufferType(IndexBuffer::IndexType::USHORT)
		.build(*mEngine);

	ib->setBuffer(*mEngine,
		IndexBuffer::BufferDescriptor(TRIANGLE_INDICES, 6, nullptr));

	Material* mat = Material::Builder()
		.package(RESOURCES_BAKEDCOLOR_DATA, RESOURCES_BAKEDCOLOR_SIZE)
		.build(*mEngine);

	utils::Entity renderable = utils::EntityManager::get().create();
	RenderableManager::Builder(1)
		.boundingBox({ { -1, -1, -1 }, { 1, 1, 1 } })
		.material(0, mat->getDefaultInstance())
		.geometry(0, RenderableManager::PrimitiveType::TRIANGLES, vb, ib, 0, 3)
		.culling(false)
		.receiveShadows(false)
		.castShadows(false)
		.build(*mEngine, renderable);

	mScene->addEntity(renderable);

	constexpr float ZOOM = 1.5f;
	const uint32_t w = mView->getViewport().width;
	const uint32_t h = mView->getViewport().height;

	const float aspect = (float)w / h;

	mCamera->setProjection(Camera::Projection::ORTHO,
		-aspect * ZOOM, aspect * ZOOM,
		-ZOOM, ZOOM, 0, 1);

	auto& tcm = mEngine->getTransformManager();
	tcm.setTransform(tcm.getInstance(renderable),
		filament::math::mat4f::rotation(0, filament::math::float3{ 0, 0, 1 }));

	closure_t print_to_png = [](uint8_t* rgba, uint32_t width, uint32_t height) {
		image::LinearImage image(image::toLinear<uint8_t>(swapchain_width, swapchain_height, swapchain_width * 3,
			static_cast<uint8_t*>(rgba)));
		std::cout << std::endl << "number of channels" << image.getChannels() << std::endl;
		std::cout << "height " << image.getHeight() << std::endl;
		std::cout << "width" << image.getWidth() << std::endl;

		std::string name = "C:/LanguagePrototyping/Filament_integration/trial_image.png";
		std::ofstream outputStream(name, std::ios::binary | std::ios::trunc);
		image::ImageEncoder::encode(outputStream, image::ImageEncoder::Format::PNG, image, "", name);
		outputStream.close();
	};

	auto callback = [](void* buffer, size_t size, void* user)
	{
		console->info("data was read");
		closure_t* closure = (closure_t*)user;
		uint8_t* rgba = (uint8_t*)buffer;
		(*closure)(rgba, swapchain_width, swapchain_height);
		delete closure;
	};

	closure_t* pointer_to_print_png = new closure_t{ std::move(print_to_png) };

	size_t size = swapchain_height * swapchain_width * 4;
	void* buffer = malloc(size);
	memset(buffer, 0, size);

	filament::backend::PixelBufferDescriptor pd{ buffer, size,
		filament::backend::PixelDataFormat::RGBA, filament::backend::PixelDataType::UBYTE,
		callback,pointer_to_print_png };

	mRenderer->renderStandaloneView(mView);
	
	console->info("data was requested");

	mRenderer->readPixels(offscreenRenderTarget,0, 0, swapchain_width, swapchain_height, std::move(pd));

	mEngine->flushAndWait();

	mEngine->destroy(renderable);
	mEngine->destroy(mat);
	mEngine->destroy(vb);
	mEngine->destroy(ib);
	mEngine->destroy(offscreenColorTexture);
	mEngine->destroy(offscreenRenderTarget);

	mEngine->destroyCameraComponent(mCameraEntity);
	utils::EntityManager::get().destroy(mCameraEntity);

	mEngine->destroy(mScene);
	mEngine->destroy(mView);
	mEngine->destroy(mSkybox);
	mEngine->destroy(mRenderer);
	Engine::destroy(&mEngine);
```
When I increase the size of the swapchain the time delay of the copy procedure increases exponentially. For a swapchain of about 500x500 pixels I get a time delay of about 100 milliseconds and for a swapchain of about 2000x2000 pixels I get a delay of 2 seconds. Although i know that the cpu and gpu must communicate between one another I was not expecting it to take so long. After going through the code of filament I found the data reshaper code in the file "https://github.com/google/filament/blob/main/filament/backend/src/vulkan/VulkanDriver.cpp"

```c++
const uint8_t* srcPixels;
    vkMapMemory(device, stagingMemory, 0, VK_WHOLE_SIZE, 0, (void**) &srcPixels);
    srcPixels += subResourceLayout.offset;

    // TODO: investigate why this Y-flip exists. This conditional seems to work with both
    // test_ReadPixels.cpp (readpixels from a normal render target with texture attachment) and
    // viewer_basic_test.cc (readpixels from an offscreen swap chain)
    const bool flipY = srcTexture ? true : false;

    if (!DataReshaper::reshapeImage(&pbd, getComponentType(srcFormat), srcPixels,
            subResourceLayout.rowPitch, width, height, swizzle, flipY)) {
        utils::slog.e << "Unsupported PixelDataFormat or PixelDataType" << utils::io::endl;
    }

    vkUnmapMemory(device, stagingMemory);

    mDisposer.createDisposable((void*)stagingImage, [=] () {
        vkDestroyImage(device, stagingImage, nullptr);
        vkFreeMemory(device, stagingMemory, nullptr);
    });

    scheduleDestroy(std::move(pbd));
```
And I realized that datareshaper has two for nested loops which should introduce some delay. Maybe it was my misunderstanding. I can try to measure elapsed time in this block of code to make sure that this does not introduce that much of a delay.

**Comment by romainguy** [comment]
@prideout Do we skip reshaping when source and destination are the same format? I'd assume RGBA/UBYTE wouldn't need reshaping.

**Comment by pixelflinger** [comment]
Also, 2s seems like a lot. Can we investigate of this is expected?

**Comment by Joaopmoliveira** [comment]
For further clarification, the 2 second (2000x2000 pixels) and 0.1 second (500x500 pixels) was for debug mode. With release mode I obtain:
(2000x2000 pixels) - 1.495 sec
(500x500 pixels) - 0.115 sec
I started to look at the other backends and they differ from one another in terms of implementing this copy procedure (the opengl one does not seem to require the datareshaper but then I don't understand how it deals with unknown formats). If you want me to, I can measure against the opengl one to see the difference

**Comment by pixelflinger** [comment]
The opengl backend doesn't have to deal with unknown formats, because by definition the formats filament supports are the ones supported by gles3.0.

The conversion, if needed, is done by the driver and performance is going to vary from one driver to another.

**Comment by Joaopmoliveira** [comment]
Understood, so in Vulkan's case it would be necessary to check the supported formats, store them on a list and check if the driver can indeed perform the conversion under the hood and if not ,call the data reshaper, thus accelerating the copy procedure in certain cases?
For my specific use case I need to render the 3D result over a swapchain which is associated with SKIA, but Filament only supports shared resources with OpenGL (according to what I read from the previous issues), so I am left with the option of copying the image by hand.  But a delay of 2 seconds is too large.. Maybe my sample contains a mistake that delays the process, but I cannot find it.

**Comment by romainguy** [comment]
"For my specific use case I need to render the 3D result over a swapchain which is associated with SKIA"

Can you explain in more detail what you need to do exactly and on which platform?

**Comment by Joaopmoliveira** [comment]
I want to target Desktops (windows mainly for now), render 2D shapes, text, etc (thus SKIA) and in some situations I need to render a 3D scene over a portion of the swapchain. 

Because Filament must be a closed system I cannot use the internal device of Filament with SKIA, thus I create a separate Vulkan device with all the necessary apparatus which I then fuse with SKIA (I need a Vulkan device, swapchain, etc...).

If it was possible to share a VkImage from SKIA, give it to Filament as a render target it would be perfect, but from the documentation I have read it seems to be impossible with Vulkan (seems possible with OpenGL).

My requirements are not heavy, so reading the pixels back, although not a perfect solution, does the trick.

**Comment by prideout** [comment]
We can improve ReadPixels but it will always be a poor solution.  Since we already have a `Texture::import()` that works with Metal and OpenGL, maybe we could make it work for Vulkan too, by allowing VkImage to be passed in and cast to intptr_t. 
 Unfortunately there would be many tricky aspects to such an API:

1) How do we manage image layout?  It might make sense for Filament to change its layout as needed, (e.g. SHADER_READ_ONLY_OPTIMAL or COLOR_ATTACHMENT_OPTIMAL).  At the very least, Filament needs to know the layout in order to set up descriptor sets properly.

2) If an external renderer (in your case, Skia) needs to render into the VkImage on every frame, we'd need a synchronization mechanism (our `Stream` API for Android already has this).  Also, would the client be responsible for inserting pipeline barriers, or would Filament?

3) When creating a VkImage, Vulkan requires you to specify a set of queue family indices that will access the image.  There needs to be some sort of agreement here between Filament and its clients.

**Comment by pixelflinger** [comment]
I think we should investigate the 2s thing though. That's seems insane to me.

**Comment by Joaopmoliveira** [comment]
Is there such handshake agreements with the other backends, or is it simpler to share contexts with them? I never used OpenGL or Metal, so I am clueless here. In my case I don't even want to render on top of the Image, so long as I could copy into into my own swapchain I would be happy.

With SKIA one can provide semaphores which let the user draw something over the image and thus achieve some sort of synchronization , maybe follow a similar logic in this case, but with higher restrictions. It depends on the level of freedom that the user code would have while interacting with the library, if the goal is to copy the rendering result of Filament, the restrictions are much simpler, if the goal is to introduce rendering over filament, it is indeed even more complex.

Assuming the simpler case, to your first point, specify the VkImage, provide the render target layout which is desired when reading from user code and a semaphore which filament could use to signal the user code when it was done rendering unto the render target. This would restrict how the user can share data with Filament in such a way that it would not mess with internal information of the engine and it would guarantee that the layout would be consistent.

But I agree with you, this goes against the design of the library from what I can tell. And to your last point, its exactly why SKIA forces the user to create all the Vulkan objects, but the drawback is exposing quite a lot of Vulkan related API. This design restriction makes it very hard to solve your third point.

With the OpenGL case, how is synchronization achieved?

**Comment by Joaopmoliveira** [comment]
@pixelflinger I will take a while but I can time the code on my machine related with the DataReshaper just to make sure that it is indeed the culprit and not another silly mistake along my code.

**Comment by prideout** [comment]
I wrote a little test that does 100 runs of ReadPixels using a 2000x2000 RGBA U8 buffer, and I'm measuring < 20 ms for each run on Vulkan, which includes the flush-and-wait between each run.

**Comment by Joaopmoliveira** [comment]
@prideout I will build the test on my PC once the merge is complete. The results clearly show that I am wrong, but I still cannot obtain a performance close you what is displayed these tests

**Comment by Joaopmoliveira** [comment]
A small question in the meanwhile the option INSTALL_BACKEND_TEST is only available on apple? although I signal the flag, the backend tests target is not generated on Visual studio

**Comment by prideout** [comment]
Yes, the backend test are still a bit janky, so far we've only used them for manual testing on mac.

**Comment by Joaopmoliveira** [comment]
Ok, I will try to adapt them so that I can run your test. I am really curious if I obtain that 18 ms performance, or if there is a problem with the compiler I am using. Or maybe I am introducing some delay due to an obscure line of code.

**Comment by bejado** [comment]
Backend tests currently only work on Apple platforms.

**Comment by prideout** [comment]
@Joaopmoliveira I think we should close this for now, please file a new issue after you narrow this down a bit, using an appropriate bug title :)
